### ISSUE TITLE
System.Net.Http.CurlException: SSL peer certificate or SSH remote key was not OK

### ISSUE THREAD

**Issue by parallelbgls** [dotnet/corefx#21429]
This exception only occurs in docker image microsoft/dotnet:1.1.2-runtime but not in microsoft/dotnet:1.1.2-runtime-nanoserver. Codes are the same.

Our company is using GeoTrust G3 WildCart sign.

**Comment by Jiayili1** [comment]
This test:
System.Net.Http.Functional.Tests.HttpClientHandler_DangerousAcceptAllCertificatesValidator_Test/SetDelegate_ConnectionSucceeds(acceptedProtocol: None, requestOnlyThisProtocol: False) 

Failed with same issue. Detail: https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20170623.01/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.HttpClientHandler_DangerousAcceptAllCertificatesValidator_Test~2FSetDelegate_ConnectionSucceeds(acceptedProtocol:%20None,%20requestOnlyThisProtocol:%20False)

Message:
~~~
System.Net.Http.HttpRequestException : An error occurred while sending the request.\n---- System.Net.Http.CurlException : SSL peer certificate or SSH remote key was not OK
~~~

Stack Trace:
~~~
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpClient.<FinishSendAsyncBuffered>d__58.MoveNext() in /Users/buildagent/agent/_work/45/s/corefx/src/System.Net.Http/src/System/Net/Http/HttpClient.cs:line 462
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.Functional.Tests.HttpClientHandler_DangerousAcceptAllCertificatesValidator_Test.<InvalidCertificateServers_CertificateValidationDisabled_Succeeds>d__3.MoveNext() in /Users/buildagent/agent/_work/45/s/corefx/src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.AcceptAllCerts.cs:line 68
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
----- Inner Stack Trace -----
   at System.Net.Http.CurlHandler.ThrowIfCURLEError(CURLcode error) in /Users/buildagent/agent/_work/45/s/corefx/src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs:line 641
   at System.Net.Http.CurlHandler.MultiAgent.FinishRequest(StrongToWeakReference`1 easyWrapper, CURLcode messageResult) in /Users/buildagent/agent/_work/45/s/corefx/src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs:line 875
~~~

**Comment by danmosemsft** [comment]
@kendrahavens what distro is microsoft/dotnet:1.1.2-runtime ?

Removing the mac label since ther'es no reason to think this is a related cause.

**Comment by kendrahavens** [comment]
@danmosemsft The linux container is debian:jessie based. The dockerfile can be found in the [dotnet-docker repo](https://github.com/dotnet/dotnet-docker/blob/master/1.1/runtime/jessie/Dockerfile). I'm not sure if this is relevant to you, but the 1.1.2-runtime tag is multi-arch so depending on whether Docker for Windows is in Windows or Linux container mode the base image will adapt though it was stated this issue wasn't present in nanoserver.

**Comment by Jiayili1** [comment]
Failed again: https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20170703.01/workItem/System.Net.Http.Functional.Tests/analysis/xunit/System.Net.Http.Functional.Tests.HttpClientHandler_DangerousAcceptAllCertificatesValidator_Test~2FInvalidCertificateServers_CertificateValidationDisabled_Succeeds(url%3A%20%5C%22https%3A~2F~2Fwrong.host.badss...

**Comment by shiitake** [comment]
I'm having the same error using the microsoft/dotnet:2.0.0-runtime and microsoft/dotnet:2.0.0-sdk images.  I'm running this in Docker for Windows (17.09.0-ce-win33 13620)

Is there any sort of known workaround?

**Comment by mbearup** [comment]
I'm hitting this same error. Do we at least know if it's restricted to Docker or a specific version of CURL?

**Comment by andremacdowell** [comment]
Same error here, dotnet core 2.0 app in a rheal container trying to upload a file to Minio using the Amazon.S3 nuget.

**Comment by hmarcelodn** [comment]
Same issue with net core 2 running this in Ubuntu todays latest version today

**Comment by karelz** [comment]
Official Helix test failures analysis:
* Test `SetDelegate_ConnectionSucceeds` failed only during 6/27-7/4 -- 4x in each run on Mac only. No repro since.
* Test `InvalidCertificateServers_CertificateValidationDisabled_Succeeds` failed only during 6/22-7/4 -- 1x per each run on Mac only. No repro since.

The Mac failures were likely unrelated to the original problem reported here, deleting the comments above to avoid clutter.

**Comment by karelz** [comment]
Can someone please provide us with a minimal repro, so tat we can try to reproduce in house? Thanks!
cc @Priya91

**Comment by karelz** [comment]
Closing as no repro. If somebody has a minimal repro, we can reopen. Thanks!

**Comment by provanguard** [comment]
I am facing the same error when calling SaltStack API from an Azure Web App with Containers support, running NET Core 2.0 Web API solution, using Flurl or RestSharp clients.

Dockerfile is using:
FROM microsoft/aspnetcore:2.0.8-jessie

```
using Flurl;
using Flurl.Http;

var result = await (saltMaster.Url + ":" + saltMaster.Port)
                    .AppendPathSegment("login")
                    .WithHeader("Accept", "application/json")
                    .PostJsonAsync(new { username = saltMaster.Username, password = saltMaster.Password, eauth = "pam" })
                    .ReceiveJson<SaltLoginResult>();
```

**Comment by karelz** [comment]
@provanguard thanks - let's follow up on your repro and how to investigate further in the new issue you created - #29891.

I am surprised to see 5 downvotes on my closing above ... if anyone else has a way to reproduce it, please speak up. Without a repro or additional information it is not actionable for us. Also note that .NET Core 2.1 does not use libcurl anymore, so trying it on 2.1 RC may get rid of the problem entirely. Let us know if you try it and what is the result. Thanks!

**Comment by HarpyWar** [comment]
Code with the following URL throws exception `SSL peer certificate or SSH remote key was not OK` on Linuxю But it works on Windows.
```c#
using (var wc = new WebClient())
{
   Console.WriteLine( wc.DownloadString("https://www.radio174stream.ru:8080") );
}
```

**Comment by karelz** [comment]
@HarpyWar can you please check if it is similar certificate problem as in #29891?

**Comment by HarpyWar** [comment]
Full exception message:
`Unhandled Exception: System.Net.WebException: An error occurred while sending the request. SSL peer certificate or SSH remote key was not OK`

dotnet version 2.1.105
Tested on clean envoronments on Debian 8 and 9.

I think a source of the error is the same, and it can be solved by replacing a certificate on a server (I don't know what exactly wrong there) or by using a solution which I wrote above (`ServicePointManager`). 
The same error message appears with other programming languages on Linux, and  it always recommended everywhere to suppress certificate errors on a client side.
